include/dfrn.php

```diff
@@ -0,0 +1,791 @@
+	xml_add_element($doc, $root, "link", "", $attributes);
+
+	$attributes = array("rel" => "alternate", "type" => "text/html", "href" => $alternatelink);
+	xml_add_element($doc, $root, "link", "", $attributes);
+
+	ostatus_hublinks($doc, $root);
+
+	if ($public) {
+		$attributes = array("rel" => "salmon", "href" => $a->get_baseurl()."/salmon/".$owner["nick"]);
+		xml_add_element($doc, $root, "link", "", $attributes);
+
+		$attributes = array("rel" => "http://salmon-protocol.org/ns/salmon-replies", "href" => $a->get_baseurl()."/salmon/".$owner["nick"]);
+		xml_add_element($doc, $root, "link", "", $attributes);
+
+		$attributes = array("rel" => "http://salmon-protocol.org/ns/salmon-mention", "href" => $a->get_baseurl()."/salmon/".$owner["nick"]);
+		xml_add_element($doc, $root, "link", "", $attributes);
+	}
+
+	if ($owner['page-flags'] == PAGE_COMMUNITY)
+		xml_add_element($doc, $root, "dfrn:community", 1);
+
+	xml_add_element($doc, $root, "updated", datetime_convert("UTC", "UTC", "now", ATOM_TIME));
+
+	$author = dfrn_add_author($doc, $owner, $authorelement, $public);
+	$root->appendChild($author);
+
+	return $root;
+}
+
+/**
+ * @brief Adds the author element in the header for the DFRN protocol
+ *
+ * @param object $doc XML document
+ * @param array $owner Owner record
+ * @param string $authorelement Element name for the author
+ *
+ * @return object XML author object
+ */
+function dfrn_add_author($doc, $owner, $authorelement, $public) {
+	$a = get_app();
+
+	$author = $doc->createElement($authorelement);
+
+	$namdate = datetime_convert('UTC', 'UTC', $owner['name-date'].'+00:00' , ATOM_TIME);
+	$uridate = datetime_convert('UTC', 'UTC', $owner['uri-date'].'+00:00', ATOM_TIME);
+	$picdate = datetime_convert('UTC', 'UTC', $owner['avatar-date'].'+00:00', ATOM_TIME);
+
+	$attributes = array("dfrn:updated" => $namdate);
+	xml_add_element($doc, $author, "name", $owner["name"], $attributes);
+
+	$attributes = array("dfrn:updated" => $namdate);
+	xml_add_element($doc, $author, "uri", $a->get_baseurl().'/profile/'.$owner["nickname"], $attributes);
+
+	$attributes = array("rel" => "photo", "type" => "image/jpeg", "dfrn:updated" => $picdate,
+				"media:width" => 175, "media:height" => 175, "href" => $owner['photo']);
+	xml_add_element($doc, $author, "link", "", $attributes);
+
+	$attributes = array("rel" => "avatar", "type" => "image/jpeg", "dfrn:updated" => $picdate,
+				"media:width" => 175, "media:height" => 175, "href" => $owner['photo']);
+	xml_add_element($doc, $author, "link", "", $attributes);
+
+	$birthday = feed_birthday($owner['user_uid'], $owner['timezone']);
+
+	if ($birthday)
+		xml_add_element($doc, $author, "dfrn:birthday", $birthday);
+
+	// The following fields will only be generated if this isn't for a public feed
+	if ($public)
+		return $author;
+
+	// Only show contact details when we are allowed to
+	$r = q("SELECT `profile`.`about`, `profile`.`name`, `profile`.`homepage`, `user`.`nickname`, `user`.`timezone`,
```

Comment:
I don't agree with that. If we always would use abstract calls instead of direct queries then I would agree. We don't even have some code logic in that query (in difference to the "get_contact_details_by_url" function). And that query would have to include many different fields from three different tables (profile, user, contact) to be complete. And last but not least I don't see many places were we would need that call.